hack/generator/pkg/jsonast/jsonast.go

```diff
@@ -463,22 +453,23 @@ func refHandler(ctx context.Context, scanner *SchemaScanner, schema *gojsonschem
 		return transformation, nil
 	}
 
-	return generateDefinitionsFor(ctx, scanner, typeName, isResource, url, schema.RefSchema)
+	return generateDefinitionsFor(ctx, scanner, typeName, schema.RefSchema)
 }
 
 func generateDefinitionsFor(
 	ctx context.Context,
 	scanner *SchemaScanner,
 	typeName astmodel.TypeName,
-	isResource bool,
-	url *url.URL,
 	schema *gojsonschema.SubSchema) (astmodel.Type, error) {
 
 	schemaType, err := getSubSchemaType(schema)
 	if err != nil {
 		return nil, err
 	}
 
+	url := schema.ID.GetUrl()
+	isResource := isResource(url)
+
```

Comment:
are you sure this change is okay? It will be getting the URL for the ref’d schema and not the schema itself

Reply:
It should be.

`generateDefinitionsFor` was called from two places:
`refHandler`, which does:
```
func refHandler... {
    url := schema.Ref.GetUrl()
    isResource := isResource(url)
    ...
    return generateDefinitionsFor(ctx, scanner, typeName, isResource, url, schema.RefSchema)
}
```

and

```
func (scanner *SchemaScanner) GenerateDefinitions ... {
    url := schema.ID.GetUrl()
    ...
    _, err = generateDefinitionsFor(ctx, scanner, rootTypeName, false, url, schema)
}
```

In both cases, the passed schema should be the same entity which the `url` was gotten from in the original caller. The second one is obviously correct. In the first one, `schema.Ref.GetUrl()` and `schema.RefSchema.ID.GetUrl()` return the same thing.

I think this change additionally makes sense -- it doesn't make sense to call `generateDefinitionsFor` with a url that _doesn't_ match the schema provided (or with `isResource` true when the schema in question is not actually pointing to a resource). Really the existing 3 parameters (`isResource`, `url`, and `schema`) are all tightly coupled and the first 2 have to have been derived from the 3rd for the call to be correct - this just makes that requirement official in the signature.

I added some sanity logging and ran the full generator just to triple check and here's a snippet of what I got:
```
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/Microsoft.Resources.json#/resourceDefinitions/links
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/Microsoft.Resources.json#/resourceDefinitions/links
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/Microsoft.Authorization.json#/resourceDefinitions/locks
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/Microsoft.Authorization.json#/resourceDefinitions/locks
 RefHandler url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/resourceDefinitions/roleAssignments
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/resourceDefinitions/roleAssignments
 RefHandler url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/RoleAssignmentProperties
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/RoleAssignmentProperties
 RefHandler url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/resourceDefinitions/roleDefinitions
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/resourceDefinitions/roleDefinitions
 RefHandler url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/RoleDefinitionProperties
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/RoleDefinitionProperties
 RefHandler url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/Permission
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2017-09-01/Microsoft.Authorization.json#/definitions/Permission
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/output
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/output
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterTypes
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterTypes
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterValueTypes
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterValueTypes
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/outputCopy
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/outputCopy
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameter
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameter
W0803 10:11:24.928653   28608 jsonast.go:714] Interface assumption unproven for https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameter
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterTypes
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterTypes
 RefHandler url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterValueTypes
 generateDefinitionsFor url: https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/parameterValueTypes
```

Reply:
Looks good then. I think I have to rollback this slightly for #205 though since the openapi types don't work in the same way, but that's something I can deal with 🙂